genie/genie-controller.go

```diff
@@ -603,3 +615,77 @@ func getK8sPodAnnotations(client *kubernetes.Clientset, k8sArgs utils.K8sArgs) (
 
 	return pod.Annotations, nil
 }
+
+func mergeWithResult(srcObj, dstObj types.Result) (types.Result, error) {
+	srcObj, err := updateRoutes(srcObj)
+	if err != nil {
+		return nil, fmt.Errorf("Routes update failed: %v", err)
+	}
+
+	if dstObj == nil {
+		return srcObj, nil
+	}
+	src, err := current.NewResultFromResult(srcObj)
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't convert old result to current version: %v", err)
+	}
+	dst, err := current.NewResultFromResult(dstObj)
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't convert old result to current version: %v", err)
+	}
+
+	ifacesLength := len(dst.Interfaces)
+
+	for _, iface := range src.Interfaces {
+		dst.Interfaces = append(dst.Interfaces, iface)
+	}
+	for _, ip := range src.IPs {
+		if ip.Interface != -1 {
+			ip.Interface += ifacesLength
+		}
+		dst.IPs = append(dst.IPs, ip)
+	}
+	for _, route := range src.Routes {
+		dst.Routes = append(dst.Routes, route)
+	}
+
+	for _, ns := range src.DNS.Nameservers {
+		dst.DNS.Nameservers = append(dst.DNS.Nameservers, ns)
+	}
+	for _, s := range src.DNS.Search {
+		dst.DNS.Search = append(dst.DNS.Search, s)
+	}
+	for _, opt := range src.DNS.Options {
+		dst.DNS.Options = append(dst.DNS.Options, opt)
+	}
+	// TODO: what about DNS.domain?
+	return dst, nil
+}
+
+func updateRoutes(rObj types.Result) (types.Result, error) {
+	result, err := current.NewResultFromResult(rObj)
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't convert old result to current version: %v", err)
+	}
+	if len(result.Routes) == 0 {
+		return result, nil
+	}
+
+	var defaultGw net.IP
+	for _, ip := range result.IPs {
+		if ip.Gateway != nil {
+			defaultGw = ip.Gateway
+			break
+		}
+	}
+
+	for _, route := range result.Routes {
+		if route.GW == nil {
+			if defaultGw == nil {
+				return nil, fmt.Errorf("Route has not set GW and the default GW is missing: %v", result)
+			}
+			route.GW = defaultGw
```

Comment:
This statement will assign the value into 'route' variable only; it will not get assigned into 'result'.

Reply:
Routes is a list of pointers to Route structure https://github.com/containernetworking/cni/blob/master/pkg/types/current/types.go#L140 it's visible in result